### PR TITLE
[CI] Add initial ci build test for XPU

### PR DESCRIPTION
**Pull request by chuanqi129**
Add initial CI build test for XPU, which will be triggered by label `ciflow/xpu` for current stage. 

Works for RFC #114850 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: .ci/docker/README.md
Comment (EikanWang) [inline .ci/docker/README.md:22]: `ubuntu-xpu`

File: .ci/docker/ubuntu-xpu/Dockerfile
Comment (EikanWang) [inline .ci/docker/ubuntu-xpu/Dockerfile:69]: For Intel GPU, we may use another file to indicate triton commit pin.

File: .ci/docker/build.sh
Comment (EikanWang) [inline .ci/docker/build.sh:230]: Compared to other backends, may I know why we do not need to define `NINJA_VERSION` and `TRITON`?

File: .ci/docker/ubuntu-xpu/Dockerfile
Comment (EikanWang) [inline .ci/docker/ubuntu-xpu/Dockerfile:81]: Since the DB is optional, do we have to add it to the docker at the moment?

File: .ci/docker/ubuntu-xpu/Dockerfile
Comment (EikanWang) [inline .ci/docker/ubuntu-xpu/Dockerfile:88]: Ditto

File: .ci/docker/ubuntu-xpu/Dockerfile
Comment (EikanWang) [inline .ci/docker/ubuntu-xpu/Dockerfile:121]: "XPU"?

File: .ci/docker/ubuntu-xpu/Dockerfile
Comment (EikanWang) [inline .ci/docker/ubuntu-xpu/Dockerfile:123]: `${XPU_VERSION}`?

File: .ci/docker/ubuntu-xpu/Dockerfile
Comment (EikanWang) [inline .ci/docker/ubuntu-xpu/Dockerfile:131]: Why do we need to copy the llvm from torch to /opt?

**Review by EikanWang — CHANGES_REQUESTED** [review]

File: .ci/docker/README.md
Comment (chuanqi129) [reply .ci/docker/README.md]: Thanks, modified typo.

File: .ci/docker/build.sh
Comment (chuanqi129) [reply .ci/docker/build.sh:231]: Add NINJA_VERSION back, but for TRITON still need wait for the triton xpu backend update to latest version.

File: .ci/docker/ubuntu-xpu/Dockerfile
Comment (chuanqi129) [reply .ci/docker/ubuntu-xpu/Dockerfile:81]: Just follow other docker images, and the original Dockerfile comes from https://github.com/pytorch/pytorch/blob/main/.ci/docker/ubuntu/Dockerfile

**Comment by EikanWang** [comment]
@chuanqi129 , please add "ciflow/xpu" to `.github/pytorch-probot.yml`

**Comment by chuanqi129** [comment]
> @chuanqi129 , please add "ciflow/xpu" to `.github/pytorch-probot.yml`

Yes, it already included in this PR. And the xpu test can be triggered now even through there is a warning message in here. Please refer: https://github.com/pytorch/pytorch/actions/runs/7271951239/job/19813286775?pr=116100

File: .ci/docker/build.sh
Comment (EikanWang) [reply .ci/docker/build.sh:231]: So, could you please add TODO?

Comment (chuanqi129) [reply .ci/docker/build.sh:231]: Added

**Review by EikanWang — APPROVED** [review]

**Comment by EikanWang** [comment]
@malfet , may I know your comments?

File: .ci/docker/ubuntu-xpu/Dockerfile
Comment (huydhn) [inline .ci/docker/ubuntu-xpu/Dockerfile:74]: Nit: I don't think ExecuTorch is needed here.  This can be removed (it doesn't cause any harm besides bloating the image)

File: .ci/docker/common/install_gcc.sh
Comment (huydhn) [inline .ci/docker/common/install_gcc.sh:7]: Nit: It makes more sense to me to just add gpg-agent into the list here https://github.com/pytorch/pytorch/blob/main/.ci/docker/common/install_base.sh#L77.  If it has been installed, it should be a no-op, right?

File: .github/workflows/xpu.yml
Comment (huydhn) [inline .github/workflows/xpu.yml:14]: Just a note that this is only building PyTorch.  I guess that we will need to add `linux.idc.xpu` runner first before any tests can be run there.  So any changes in `test.sh` doesn't do anything yet.

**Review by huydhn — APPROVED** [review]
I have some minor comments, but overall LGTM!

File: .ci/docker/build.sh
Comment (atalman) [inline .ci/docker/build.sh:223]: You will also need to add a line in 
```
https://github.com/pytorch/pytorch/blob/main/.github/workflows/docker-builds.yml#L38
```
In order to make sure docker build is working

File: .github/workflows/xpu.yml
Comment (atalman) [inline .github/workflows/xpu.yml:20]: Runner type should probably be ``linux.idc.xpu``

**Review by atalman — CHANGES_REQUESTED** [review]

File: .ci/docker/ubuntu-xpu/Dockerfile
Comment (chuanqi129) [reply .ci/docker/ubuntu-xpu/Dockerfile:74]: Yes, removed this part

File: .ci/docker/common/install_gcc.sh
Comment (chuanqi129) [reply .ci/docker/common/install_gcc.sh]: right, agree with you, move the gpg-agent installation into base script

File: .github/workflows/xpu.yml
Comment (chuanqi129) [reply .github/workflows/xpu.yml:14]: Yes, just worked with @atalman to add the xpu runners

File: .ci/docker/build.sh
Comment (chuanqi129) [reply .ci/docker/build.sh:223]: Thanks, added

File: .github/workflows/xpu.yml
Comment (chuanqi129) [reply .github/workflows/xpu.yml:20]: Sorry for the misleading in RFC, the xpu build also can use aws instance directly.

File: .ci/docker/build.sh
Comment (huydhn) [reply .ci/docker/build.sh:223]: Ah, now this explained why I did see the usual missing ECR entry error when pushing a brand new docker image name.  I guess Andrey has added it https://us-east-1.console.aws.amazon.com/ecr/repositories/private/308535385114/pytorch/pytorch-linux-jammy-xpu-2024.0-py3?region=us-east-1, so retry should work now.

Comment (chuanqi129) [reply .ci/docker/build.sh:223]: Yes, he helped to do it.

**Comment by chuanqi129** [comment]
Hi @atalman, all checks passed, please help to approve and merge if there is no more comment😁

**Review by atalman — APPROVED** [review]

**Comment by malfet** [comment]
This PR broke everything, though any PR that would touch docker will do the same right now

**Comment by chuanqi129** [comment]
> This PR broke everything, though any PR that would touch docker will do the same right now and alas revert is not a solution: https://hud.pytorch.org/pytorch/pytorch/commit/961fbbe967457da152a46b978aa1a427f18d7be7
> 
> Hopefully this will fix it: #116549

Because this PR modified `.ci/docker`, all docker images need to be rebuilt, and then the corruption problem occurs, right?

**Pull request closed without merging**
